python_modules/jupedsim/jupedsim/distributions.py

```diff
@@ -1,2 +1,1012 @@
+    distances = []
+    n = len(polygon)
+    i = 0
+    while True:
+        # find the number of the next line segment
+        following = (i + 1) % n
+        distances.append(distance_to_segment(polygon[i], polygon[following], pt))
+        i = following
+        if i == 0:
+            break
+    return min(distances)
+
+
+def distance_to_segment(a, b, e):
+    """returns the minimal distance between point e and the line segment from a to b"""
+    # vector AB
+    ab = [None, None]
+    ab[0] = b[0] - a[0]
+    ab[1] = b[1] - a[1]
+
+    # vector BP
+    be = [None, None]
+    be[0] = e[0] - b[0]
+    be[1] = e[1] - b[1]
+
+    # vector AP
+    ae = [None, None]
+    ae[0] = e[0] - a[0]
+    ae[1] = e[1] - a[1]
+
+    # Variables to store dot product
+
+    # Calculating the dot product
+    ab_be = ab[0] * be[0] + ab[1] * be[1]
+    ab_ae = ab[0] * ae[0] + ab[1] * ae[1]
+
+    # Minimum distance from
+    # point E to the line segment
+
+    # Case 1
+    if ab_be > 0:
+
+        # Finding the magnitude
+        y = e[1] - b[1]
+        x = e[0] - b[0]
+        req_ans = sqrt(x * x + y * y)
+
+    # Case 2
+    elif ab_ae < 0:
+        y = e[1] - a[1]
+        x = e[0] - a[0]
+        req_ans = sqrt(x * x + y * y)
+
+    # Case 3
+    else:
+
+        # Finding the perpendicular distance
+        x1 = ab[0]
+        y1 = ab[1]
+        x2 = ae[0]
+        y2 = ae[1]
+        mod = sqrt(x1 * x1 + y1 * y1)
+        req_ans = abs(x1 * y2 - y1 * x2) / mod
+
+    return req_ans
+
+
+def distance_between(pt1, pt2):
+    """returns the distance between point1 and point2"""
+    dx = pt2[0] - pt1[0]
+    dy = pt2[1] - pt1[1]
+    return sqrt(dx ** 2 + dy ** 2)
+
+
+def create_random_points(polygon, count, agent_radius, wall_distance, seed=None, obstacles=None, density=None):
+    """returns points randomly placed inside the polygon
+
+        :param polygon: List of corner points given as tuples
+        :param count: number of points placed
+        :param density: select a density: number of agents will be calculated and param count will not be used
+        :param agent_radius: minimal distance between points
+        :param wall_distance: minimal distance between points and the polygon
+        :param seed: define a seed for random generation
+        :param obstacles: list of polygons, no points can be inside the obstacle
+        :return: list of created points
+        if more that 10000 tries are needed to placed a valid point an Exception will be thrown
+    """
+    if obstacles is None:
+        obstacles = []
+    if seed is not None:
+        np.random.seed(seed)
+    if density is not None:
+        area = area_of_polygon(polygon)
+        # it is expected that the entire obstacle intersects with the area
+        for obstacle in obstacles:
+            area -= area_of_polygon(obstacle)
+        count = round(density * area)
+
+    # creates a grid
```

Comment:
Please extract the 2d-grid for fast 2d neighbour lookup into its own data structure with tests.

Reply:
done